internal/clients/content_sources/util.go

```diff
@@ -2,15 +2,24 @@ package content_sources
 
 import (
 	"fmt"
+	"net/url"
 )
 
-func GetBaseURL(repo ApiRepositoryResponse) (string, error) {
+func GetBaseURL(repo ApiRepositoryResponse, csReposURL *url.URL) (string, error) {
 	if repo.Origin == nil {
 		return "", fmt.Errorf("unable to read origin from repository %s", *repo.Uuid)
 	}
 	switch *repo.Origin {
 	case "upload":
-		return *repo.LatestSnapshotUrl, nil
+		if csReposURL == nil {
+			return "", fmt.Errorf("upload repositories require a content sources URL")
```

Comment:
Nit pick: returning dynamic errors lead to inability to unwrap them and check for values or types. AWS SDK does this and the end result is parsing error strings to find root cause. This simple trick is idiomatic Go to solve this:

```
// exported error value
var ErrMissingURL = errors.New("upload repositories require a content sources URL")

// ...
return "", ErrMissingURL
```

Reply:
I agree, but here the err value isn't checked, just the fact that there is an error. And I think this are all handled as 500s by default.